src/kernel/pgtbl.c

```diff
@@ -4,184 +4,36 @@
-	assert(!(orig_v & PGTBL_QUIESCENCE));
-
-	*kern_addr = (unsigned long)chal_pa2va((paddr_t)(orig_v & PGTBL_FRAME_MASK));
-	new_v      = orig_v | PGTBL_COSKMEM;
-
-	/* pa2va (value in *kern_addr) will return NULL if the page is
-	 * not kernel accessible */
-	if (unlikely(!*kern_addr)) return -EINVAL; /* cannot retype a non-kernel accessible page */
-	if (unlikely(retypetbl_kern_ref((void *)(new_v & PGTBL_FRAME_MASK)))) return -EFAULT;
-
-	/* We keep the cos_frame entry, but mark it as COSKMEM so that
-	 * we won't use it for other kernel objects. */
-	if (unlikely(cos_cas((unsigned long *)pte, orig_v, new_v) != CAS_SUCCESS)) {
-		/* restore the ref cnt. */
-		retypetbl_deref((void *)(orig_v & PGTBL_FRAME_MASK));
-		return -ECASFAIL;
-	}
-	/* Return the pte ptr, so that we can release the page if the
-	 * kobj activation failed later. */
-	*pte_ret = (unsigned long *)pte;
-
-	return 0;
+    return chal_pgtbl_kmem_act(pt, addr, kern_addr, pte_ret);
 }
 
 /* Return 1 if quiescent past since input timestamp. 0 if not. */
 int
 tlb_quiescence_check(u64_t timestamp)
 {
-	int i, quiescent = 1;
-
-	/* Did timer interrupt (which does tlb flush
-	 * periodically) happen after unmap? The periodic
-	 * flush happens on all cpus, thus only need to check
-	 * the time stamp of the current core for that case
-	 * (assuming consistent time stamp counters). */
-	if (timestamp > tlb_quiescence[get_cpuid()].last_periodic_flush) {
-		/* If no periodic flush done yet, did the
-		 * mandatory flush happen on all cores? */
-		for (i = 0; i < NUM_CPU_COS; i++) {
-			if (timestamp > tlb_quiescence[i].last_mandatory_flush) {
-				/* no go */
-				quiescent = 0;
-				break;
-			}
-		}
-	}
-	if (quiescent == 0) {
-		printk("from cpu %d, t %llu: cpu %d last mandatory flush: %llu\n", get_cpuid(), timestamp, i,
-		       tlb_quiescence[i].last_mandatory_flush);
-		for (i = 0; i < NUM_CPU_COS; i++) {
-			printk("cpu %d: flush %llu\n", i, tlb_quiescence[i].last_mandatory_flush);
-		}
-	}
-
-	return quiescent;
+    return chal_tlb_quiescence_check(timestamp);
 }
 
 int
 cap_memactivate(struct captbl *ct, struct cap_pgtbl *pt, capid_t frame_cap, capid_t dest_pt, vaddr_t vaddr)
 {
-	unsigned long *    pte, cosframe, orig_v;
-	struct cap_header *dest_pt_h;
-	u32_t              flags;
-	int                ret;
-
-	if (unlikely(pt->lvl || (pt->refcnt_flags & CAP_MEM_FROZEN_FLAG))) return -EINVAL;
-
-	dest_pt_h = captbl_lkup(ct, dest_pt);
-	if (dest_pt_h->type != CAP_PGTBL) return -EINVAL;
-	if (((struct cap_pgtbl *)dest_pt_h)->lvl) return -EINVAL;
-
-	pte = pgtbl_lkup_pte(pt->pgtbl, frame_cap, &flags);
-	if (!pte) return -EINVAL;
-	orig_v = *pte;
-
-	if (!(orig_v & PGTBL_COSFRAME) || (orig_v & PGTBL_COSKMEM)) return -EPERM;
-
-	assert(!(orig_v & PGTBL_QUIESCENCE));
-	cosframe = orig_v & PGTBL_FRAME_MASK;
-
-	ret = pgtbl_mapping_add(((struct cap_pgtbl *)dest_pt_h)->pgtbl, vaddr, cosframe, PGTBL_USER_DEF);
-
-	return ret;
```

Comment:
Why should this function be in ```chal```? This code seems rather generic to me.
Should ```pgtbl_mapping_add``` be also in ```chal``` or not?

Reply:
The page table is largely in chal. The code logic is generic but I need to add a few parameters to it. These parameters will appear in the following commits.